server: disable DistSQL in server's InternalExecutor until migrations complete

asubiotto opened the pull request:
DistSQL is fragile on node startup (see #44101) and is therefore disabled in
the migration manager's internal executor. However, migrations sometimes have
side effects and sometimes end up using the SQL server's internal executor
which had DistSQL enabled.

This commit disabled DistSQL for this internal executor until the migrations
complete to avoid flakiness.

Release note: None (testing flake fix)

Fixes #50000

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51053)
<!-- Reviewable:end -->

andreimatei commented on pkg/server/server_sql.go:
```diff
@@ -610,6 +610,23 @@ func (s *sqlServer) start(
 			// present situation).
 			DistSQLMode: sessiondata.DistSQLOff,
 		})
+	{
```
I guess I'd get this of the `migrationsExecutor`.

asubiotto replied:
I'm not sure I understand this comment.

andreimatei commented on pkg/server/server_sql.go:
```diff
@@ -610,6 +610,23 @@ func (s *sqlServer) start(
+		// The server's internalExecutor might be used as a side effect of
+		// migrations, so disable distribution for the same reason as above until
+		// all migrations have been run.
+		sessionDataCopy, sessionDataWasNil := s.internalExecutor.GetSessionData()
```
Can the results of `GetSessionData()` vary at all? If not, let's just assert that they are what they should be and don't bother keeping track of the old settings. In fact, I wouldn't even call `GetSessionData()`.

asubiotto replied:
I don't think so since this is all in the startup path. Removed `GetSessionData`.

andreimatei commented on pkg/sql/internal.go:
```diff
@@ -93,14 +93,27 @@ func MakeInternalExecutor(
 // SetSessionData cannot be called concurently with query execution.
 func (ie *InternalExecutor) SetSessionData(sessionData *sessiondata.SessionData) {
-	ie.s.populateMinimalSessionData(sessionData)
+	if sessionData != nil {
```
I don't think this guy should accept a nil. I hope this becomes moot if it's true that `GetSessionData()` always returns `nil` in `Server.Start`.

asubiotto replied:
Removed.

asubiotto commented:
Look's like the internal executor is already in use at this point (`testrace` failure). I added a mutex around `sessionData` which I think is ok because it's not a hot path.

Something I'm confused about is how exactly an internal executor uses `DistSQL`? It doesn't look like anywhere sets an explicit sessiondata with distsql enabled. An empty sessiondata is equivalent to having distsql turned off right?

Commit:
server: disable DistSQL in server's InternalExecutor until migrations complete

DistSQL is fragile on node startup (see #44101) and is therefore disabled in
the migration manager's internal executor. However, migrations sometimes have
side effects and sometimes end up using the SQL server's internal executor
which had DistSQL enabled.

This commit disabled DistSQL for this internal executor until the migrations
complete to avoid flakiness.

Release note: None (testing flake fix)

asubiotto commented:
Planning to add a test to this.

andreimatei reviewed (commented):
> Something I'm confused about is how exactly an internal executor uses DistSQL? It doesn't look like anywhere sets an explicit sessiondata with distsql enabled. An empty sessiondata is equivalent to having distsql turned off right?


Good question. I've just tested it and the following results in `off`.
```
func TestXXX(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)

	ctx := context.Background()
	params, _ := tests.CreateTestServerParams()
	s, _, _ := serverutils.StartServer(t, params)
	defer s.Stopper().Stop(ctx)

	ie := s.InternalExecutor().(*sql.InternalExecutor)
	row, err := ie.QueryRowEx(ctx, "test", nil, /* txn */
		sqlbase.InternalExecutorSessionDataOverride{User: security.RootUser},
		"SHOW DISTSQL")
	if err != nil {
		t.Fatal(err)
	}
	if len(row) != 1 {
		t.Fatalf("expected 1 col, got: %d", len(row))
	}
	r, ok := row[0].(*tree.DString)
	log.Infof(ctx, "!!! ok: %t. r: %s", ok, r)
}
```

I think before #47019 maybe the session vars for an internal executor were initialized from the `GlobalsDefaults`, but not now (and I don't think that was generally the case then either).

Let's figure this out. Can you detail what seems to happen in #50000 ?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51053#-:-MBo3D_yARMT22cQf3yk:b-n06134)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)

___
*[pkg/server/server_sql.go, line 613 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51053#-MBciN2hdlm5vzHQJV6x-r1-613:-MBnsGY0D1dvSzf-FpBs:bz669h) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5db9ce389c84f4fe57c7ceb9762b57aac48145/pkg/server/server_sql.go#L613)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm not sure I understand this comment.
</blockquote></details>

You can just have `migMgr` use `s.internalExecutor`, can you not? There's no longer a point to `migrationsExecutor`, is there?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
I should have said that uses of the internal executor from normal user sessions inherit the parent session's data through a `SetSessionData()` - which does get its original values from those `GlobalDefaults`, but the migrations start with empty data, so there should be nothing for their children to inherit I would think.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51053#-:-MBo5ETuChCG_S-PskUu:b-ubug6e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)



<!-- Sent from Reviewable.io -->

asubiotto commented:
Won't be able to get to this until tomorrow. In #50000, some migration calls `GetAllRoles` which uses the executor in the planner, which is the `internalExecutor` in this PR. This was the comment on the original issue: https://github.com/cockroachdb/cockroach/issues/47024#issuecomment-612980142

knz commented:
Is there a way to move forward on this, viz failing roachtests.

asubiotto commented:
Checked the distsql version on a manual run forcing migrations and it seems that distsql is off so don't think this PR would change anything. Guess it's back to the drawing board. Recent runs of `acceptance/version-upgrade` with the `no inbound stream connection` don't seem to have artifacts (e.g. https://teamcity.cockroachdb.com/viewLog.html?buildId=2119855&tab=buildLog). @knz do you by any chance have a link to one that does? In the meantime will `roachprod-stress` with some more instrumentation.

knz commented:
> @knz do you by any chance have a link to one that does? 

not at this time no unfortunately. But i'll be on the lookout.

> In the meantime will roachprod-stress with some more instrumentation.

Yes that's what I'd do too.

asubiotto commented:
Superseded by #52624
